### ISSUE TITLE
observable map modifications outside action do not rise errors with useStrict

### ISSUE THREAD

**Issue by fedorl** [mobxjs/mobx#940]
`map.set(existingKey, x)` in the strict mode outside an action fails as expected with `Invariant failed`, requiring an action context.

However, map.set(nonExistentKey, x) as well as other modifying operations e.g.  map.delete(), map.clear() silently run without any error even when no action is present. 

**Any** operation that modifies anything in the observable map state outside of an action must fail; the same way as an existing key replacement and the same way as ObservableArray.clear() or ObservableArray.push().

Current behavior is not consistent and loses the value of useStrict for maps, because one can't rely on it anymore. Consider two equivalent operations `set(key, v); set(key, v)` and `set(key);delete(key);set(key); ` which have now completely different outcome, one will fail and one will work outside of an action. 

Not only does it allow uncontrolled state mutations, it does it in an unpredictable way depending on other mutations and their order. 

`
class PropTest {
  @observable propMap = observable.map();
}

const key = 'testKey';
autorun((r) => {
   const val = propTest.propMap.get(key);
   console.info('------> ' + val);
});

propTest.propMap.set(key, 1);
propTest.propMap.set(key, 2); // -> Invariant failed as expected

//below runs fine which is not what one would expect in strict mode
propTest.propMap.set(key, 1);
propTest.propMap.delete(key);
propTest.propMap.set(key, 2);
propTest.propMap.clear();
`

If there are implementation difficulties, this limitation should be reflected in documentation, as this inconsistency seems to be an alarming issue.

**Comment by urugator** [comment]
I can confirm that only value modifications of existing keys throw, key modifications (introducion of new key, `delete`, `clear`) doesn't throw. 
[Here is the fiddle to check the behavior](http://jsfiddle.net/kwp9jqfm/14/)

**Comment by mweststrate** [comment]
Confirmed, this is probably caused by the fact that map uses actions internally, https://github.com/mobxjs/mobx/blob/master/src/types/observablemap.ts#L198. 

Either before running the internal action, `checkIfStateModificationsAreAllowed()`, should be called, or `runInTransaction` should not start an action (but still do the batching!).

PR's are welcome :)

**Comment by roperzh** [comment]
Hey @mweststrate , sorry to bother! I've been looking at this and I wanted to ask before proceeding. 

Since `checkIfStateModificationsAreAllowed` expects an `Atom` as an argument, should we instantiate a new `Atom` when the map is created? I'm thinking in something in the lines of [`ObservableArray`](https://github.com/mobxjs/mobx/blob/master/src/types/observablearray.ts#L102).

Thank you!

**Comment by urugator** [comment]
I may be wrong, but I think the idea is to perform the check on the map's internal observable structures whenever they are being modified in `runInTransaction` (which provides batching, but unfortunately supresses the strict mode).
To obtain an atom from observable you can use `extras.getAtom(observable)`.
So for example [here](https://github.com/mobxjs/mobx/blob/master/src/types/observablemap.ts#L149), the observable `this._keys` is being modified so we must check if it can be modified, before the action is invoked:
```javascript
checkIfStateModificationsAreAllowed(getAtom(this._keys)) // throws in strict
runInAction(() => {
  this._keys.remove(key);
})

```
Similary [here](https://github.com/mobxjs/mobx/blob/master/src/types/observablemap.ts#L168) the observable `entry` is being modified, so the same check needs to be performed. The problem is, you still have to push the check before `runInAction`, which is outside of this nested method...

So the second idea is to replace `runInAction` with a similar function, which doesn't call [`allowStateChangesStart`](https://github.com/mobxjs/mobx/blob/1cc4f5173f6da74a447c94615a62fb9c1380b397/src/core/action.ts#L94) and [`allowStateChangesEnd`](https://github.com/mobxjs/mobx/blob/1cc4f5173f6da74a447c94615a62fb9c1380b397/src/core/action.ts#L71), but still calls [`startBatch`](https://github.com/mobxjs/mobx/blob/1cc4f5173f6da74a447c94615a62fb9c1380b397/src/core/action.ts#L60) and [`endBatch`](https://github.com/mobxjs/mobx/blob/1cc4f5173f6da74a447c94615a62fb9c1380b397/src/core/action.ts#L72). (it may be a bit more complicated, dunno)
Which seems useful for cases like this and also like a better/cleaner solution.

**Comment by mweststrate** [comment]
The issue is I think simply that the methods on observable map are marked as `@action`, where `runInTransaction` should be used. The main thing withholding fixing this bug is that it will probably be a breaking change for a lot of people if they modified a map in some code that is not an action yet, that worked so far, and not anymore after fixing this.

**Comment by smikula** [comment]
I see this is fixed -- any plans for when a release including this fix will come out?

I'm glad to see that `transaction` is sticking around.  It's really helpful in [my project](https://github.com/microsoft/satcheljs) to be able to distinguish a transaction from an operation that can modify state.

**Comment by mweststrate** [comment]
Just released as 3.3.1 😊
